src/Backend/PowsyblBackend.py

```diff
@@ -0,0 +1,1540 @@
+                    l_id,
+                    l_id - self.__nb_powerline,
+                    3,
+                )
+        for l_id, pos_big_topo in enumerate(self.line_ex_pos_topo_vect):
+            if l_id < self.__nb_powerline:
+                self._big_topo_to_backend[pos_big_topo] = (l_id, l_id, 4)
+            else:
+                self._big_topo_to_backend[pos_big_topo] = (
+                    l_id,
+                    l_id - self.__nb_powerline,
+                    5,
+                )
+
+        self.theta_or = np.full(self.n_line, fill_value=np.NaN, dtype=dt_float)
+        self.theta_ex = np.full(self.n_line, fill_value=np.NaN, dtype=dt_float)
+        self.load_theta = np.full(self.n_load, fill_value=np.NaN, dtype=dt_float)
+        self.gen_theta = np.full(self.n_gen, fill_value=np.NaN, dtype=dt_float)
+        self.storage_theta = np.full(self.n_storage, fill_value=np.NaN, dtype=dt_float)
+
+        self.dim_topo = np.sum(self.sub_info)
+        self._compute_pos_big_topo()
+
+        self.line_status[:] = self._get_line_status()
+        self._topo_vect = self._get_topo_vect()
+        self.tol = 1e-5  # this is NOT the pandapower tolerance !!!! this is used to check if a storage unit
+        # produce / absorbs anything
+
+        # Create a deep copy of itself in the initial state
+        # Store it under super private attribute
+        with warnings.catch_warnings():
+            # raised on some versions of pandapower / pandas
+            warnings.simplefilter("ignore", FutureWarning)
+            self.__pp_backend_initial_grid = self._grid.deepcopy()  # will be initialized in the "assert_grid_correct"
+
+        self.map_sub_invert = {v: k for k, v in self.map_sub.items()}
+
+    def storage_deact_for_backward_comaptibility(self):
+        self._init_private_attrs()
+
+    def apply_action(self, backendAction=None):
+        """
+        INTERNAL
+
+        .. warning:: /!\\\\ Internal, do not use unless you know what you are doing /!\\\\
+
+        Specific implementation of the method to apply an action modifying a powergrid in the pypowsybl format.
+        """
+
+        if backendAction is None:
+            return
+        cls = type(self)
+
+        (
+            active_bus,
+            (prod_p, prod_v, load_p, load_q, storage),
+            topo__,
+            shunts__,
+        ) = backendAction()
+
+        # TODO Normally we don't have to handle this for the backend because inactive buses will not appear in
+        # TODO get_buses() fct for pypowsybl
+
+        # handle bus status
+        # bus_is = self._grid.get_buses()
+        # for i, (bus1_status, bus2_status) in enumerate(active_bus):
+        #     bus_is[i] = bus1_status  # no iloc for bus, don't ask me why please :-/
+        #     bus_is[i + self.__nb_bus_before] = bus2_status
+
+        if np.any(prod_p.changed):
+            self._grid.update_generators(id=self.name_gen, target_p=prod_p.values)
```

Comment:
No, you should not read the prod_p.values if the flag prod_p.changed is `False` 

This is again explained in the doc about backend creation, the provided example and the EducPandaPowerBackend...

Reply:
Correct code is:


```python
        if np.any(prod_p.changed):
            self._grid.update_generators(id=self.name_gen[prod_p.changed], target_p=prod_p.values[prod_p.changed])
```

Depending on what does the action, some elements might not be changed between consecutive steps. And in this case the associated `values` is garbage (unintialized or nan or inf or something)